### PR TITLE
Finish specification of the lifetime elision rules

### PR DESCRIPTION
**Pull request by jonas-schievink**
Closes https://github.com/ferrocene/specification/issues/84
Closes https://github.com/ferrocene/specification/issues/80

### CODE REVIEW

File: src/glossary.rst [hunk 1]
```diff
@@ -3521,7 +3531,9 @@ lifetime elision
 ^^^^^^^^^^^^^^^^
 
 :dp:`fls_dq5wkd61ry3l`
-:dt:`Lifetime elision` is a set of relaxations on the use of :t:`[lifetime]s`.
+:dt:`Lifetime elision` is a set of rules that automatically insert :t:`[lifetime
+parameter]s` and/or :t:`[lifetime argument]s` when they are elided in the source
+code.
```
Comment (kirtchev-adacore) [inline src/glossary.rst:3536]:
```suggestion
:dt:`Lifetime elision` is a set of rules that automatically insert
:t:`[lifetime parameter]s` and/or :t:`[lifetime argument]s` when they are
elided in the source code.
```
Not splitting terms that consist of multiple words across multiple lines makes it easier to grep. I went through all the chapters and ensured that none of the terms are split.

File: src/types-and-traits.rst [hunk 2]
```diff
@@ -2778,7 +2778,35 @@ Lifetime Elision
 .. rubric:: Legality Rules
 
 :dp:`fls_9wtuclhm7yz5`
-:t:`Lifetime elision` is a set of relaxations on the use of :t:`[lifetime]s`.
+:t:`Lifetime elision` is a set of rules that automatically insert :t:`[lifetime
+parameter]s` and/or :t:`[lifetime argument]s` when they are elided in the source
+code.
```
Comment (kirtchev-adacore) [inline src/types-and-traits.rst:2783]:
```suggestion
:t:`Lifetime elision` is a set of rules that automatically insert
:t:`[lifetime parameter]s` and/or :t:`[lifetime argument]s` when they are
elided in the source code.
```

File: src/types-and-traits.rst [hunk 3]
```diff
@@ -2778,7 +2778,35 @@ Lifetime Elision
...
+
+:dp:`fls_YmUQ8ZiQuycp`
+A :t:`lifetime` is elided implicitly if it is absent.
+
+:dp:`fls_sIMN6Sd8xUZz`
+:t:`Lifetime elision` rules are introduced by certain :t:`[construct]s` and may
+be nested.
+
+:dp:`fls_dIyisjNIx9dC`
+If an elided :t:`lifetime` is encountered, the innermost set of :t:`lifetime
```
Comment (kirtchev-adacore) [inline src/types-and-traits.rst:2799]: What/which are these innermost (vs outermost?) set of rules?

File: src/types-and-traits.rst [hunk 4]
```diff
@@ -2778,7 +2778,35 @@ Lifetime Elision
...
+elision` rules applies.
+
+:dp:`fls_cD0ZYi23VqWg`
+It is a static error to elide a :t:`lifetime` in a position where no
+:t:`lifetime elision` rules are active.
+
+:dp:`fls_sA4Lqc5o6cX3`
+:t:`[Lifetime]s` cannot be implicitly elided within :t:`[impl trait type]s`:
+If no :t:`lifetime bound` is present, the :t:`impl trait type` is not considered
+to be bound by any lifetime.
```
Comment (kirtchev-adacore) [inline src/types-and-traits.rst:2809]:
```suggestion
:t:`[Lifetime]s` cannot be implicitly elided within :t:`[impl trait type]s`.
If no :t:`lifetime bound` is present, the :t:`impl trait type` is not considered
to be bound by any :t:`lifetime`.
```

File: src/types-and-traits.rst [hunk 5]
```diff
@@ -2964,3 +2992,25 @@ its :t:`lifetime` :t:`elided` form is
...
+
+:dp:`fls_3p5BdLn3JbKz`
+The :t:`impl header lifetime elision` rules are as follows:
+
+* :dp:`fls_mWxIQ5wagMpa`
+  An implicitly elided :t:`lifetime` is a static error.
+
+* :dp:`fls_PfS5AlkN6ANl`
+  Every explicitly elided :t:`lifetime` is replaced with a new :t:`lifetime
+  parameter` defined on the :t:`implementation`.
```
Comment (kirtchev-adacore) [inline src/types-and-traits.rst:3016]:
```suggestion
  Every explicitly elided :t:`lifetime` is replaced with a new
  :t:`lifetime parameter` defined on the :t:`implementation`.
```

**Review by kirtchev-adacore — CHANGES_REQUESTED** [review]

File: src/types-and-traits.rst (hunk 3 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2799]: Ouf, I see what you mean. How about something long the lines of

An elided lifetime is subject to the lifetime elision rules of the narrowest context where the lifetime appears.

Or something along those lines. It is still somewhat unclear what "narrowest" means. How about

An elided lifetime is subject to the lifetime elision rules of the innermost construct where the lifetime appears.

I also don't like "appears". Any improvements on these suggestions?

**Review by kirtchev-adacore — CHANGES_REQUESTED** [review]

File: src/types-and-traits.rst (hunk 3 above)
Comment (jonas-schievink) [reply src/types-and-traits.rst]: Okay, I've rewritten this sentence to hopefully be clearer

**Review by kirtchev-adacore — APPROVED** [review]
I like the wording!

**Comment by kirtchev-adacore** [comment]
bors merge

**Pull request merged**